src/DocBlock/TypeExpression.php

```diff
@@ -267,58 +263,78 @@ public function getTypesGlue(): string
     }
 
     /**
-     * @param \Closure(self): void $callback
+     * @param \Closure(self): self $callback
      */
-    public function walkTypes(\Closure $callback): void
+    public function mapTypes(\Closure $callback): self
     {
-        $innerValueOrig = $this->value;
-
+        $value = $this->value;
         $startIndexOffset = 0;
 
         foreach ($this->innerTypeExpressions as [
             'start_index' => $startIndexOrig,
             'expression' => $inner,
         ]) {
-            $innerLengthOrig = \strlen($inner->toString());
+            $innerValueOrig = $inner->value;
 
-            $inner->walkTypes($callback);
+            $inner = $inner->mapTypes($callback);
 
-            $this->value = substr_replace(
-                $this->value,
-                $inner->toString(),
-                $startIndexOrig + $startIndexOffset,
-                $innerLengthOrig
-            );
+            if ($inner->value !== $innerValueOrig) {
+                $value = substr_replace(
+                    $value,
+                    $inner->value,
+                    $startIndexOrig + $startIndexOffset,
+                    \strlen($innerValueOrig)
+                );
 
-            $startIndexOffset += \strlen($inner->toString()) - $innerLengthOrig;
+                $startIndexOffset += \strlen($inner->value) - \strlen($innerValueOrig);
+            }
         }
 
-        $callback($this);
+        $type = $value === $this->value
+            ? $this
+            : $this->inner($value);
 
-        if ($this->value !== $innerValueOrig) {
-            $this->isUnionType = false;
-            $this->typesGlue = '|';
-            $this->innerTypeExpressions = [];
+        return $callback($type);
+    }
 
-            $this->parse();
-        }
+    /**
+     * @param \Closure(self): void $callback
+     */
+    public function walkTypes(\Closure $callback): void
```

Comment:
Why do we keep `walkTypes` function?

Reply:
Because you might want to analyse type without transforming it.

Reply:
Can you give an example? I can't think of a situation when we want to do it. We call a `mapTypes` and compare it to see if the expression will change.

Reply:
We do? I cannot find it in the code.

Reply:
IDK, but walk/map are basic operations on node tree so I belive we should have both.

Reply:
I'd say let's add when needed.

Reply:
`TypeExpression::walkTypes()` method was already present and it is public, I am not going to remove it as it will imply BC break.